### ISSUE TITLE
My benchmark for multi-threaded writes

### ISSUE THREAD

**Issue by boxerab** [axboe/liburing#571]
I've created a [benchmark program](https://github.com/GrokImageCompression/uring_mt_test) to test my particular work flow: multi-threaded writes to generate a TIFF file. This is a cmake project that is easy to build and run.

The benchmark divides an image into M strips and uses a thread pool to do some work on each strip and then write the strip to libtiff. Each thread manages its own ring - the work queue is shared. For comparison, the benchmark also supports  synchronous writes.

Here are some outputs from `/usr/bin/time` for various IO configurations for writing a 55000x88000 TIFF file on the following system :

Fedora 36 desktop : 5.17 kernel, `xfs` file system,  48 thread Threadripper with PCI4 NVME SSD, and latest liburing.

## URING

1. `uring` write with `direct` IO (2 threads)
```
Run with concurrency = 2, store to disk = 1, direct = 1, use uring = 1
 : 714.831964 ms
	Command being timed: "./iobench -c 2 -d -f /run/media/xfs/io_out.tif"
	User time (seconds): 0.91
	System time (seconds): 0.60
	Percent of CPU this job got: 194%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 0:00.78
	Average shared text size (kbytes): 0
	Average unshared data size (kbytes): 0
	Average stack size (kbytes): 0
	Average total size (kbytes): 0
	Maximum resident set size (kbytes): 741164
	Average resident set size (kbytes): 0
	Major (requiring I/O) page faults: 0
	Minor (reclaiming a frame) page faults: 184445
	Voluntary context switches: 3282
	Involuntary context switches: 7
	Swaps: 0
	File system inputs: 25
	File system outputs: 5500888
	Socket messages sent: 0
	Socket messages received: 0
	Signals delivered: 0
	Page size (bytes): 4096
	Exit status: 0
```

2. `uring` write with `direct` IO (48 threads)

```
Run with concurrency = 48, store to disk = 1, direct = 1, use uring = 1
 : 892.141297 ms
	Command being timed: "./iobench -c 0 -d -f /run/media/xfs/out.tif"
	User time (seconds): 1.92
	System time (seconds): 6.98
	Percent of CPU this job got: 915%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 0:00.97
	Average shared text size (kbytes): 0
	Average unshared data size (kbytes): 0
	Average stack size (kbytes): 0
	Average total size (kbytes): 0
	Maximum resident set size (kbytes): 1367200
	Average resident set size (kbytes): 0
	Major (requiring I/O) page faults: 2
	Minor (reclaiming a frame) page faults: 342316
	Voluntary context switches: 184711
	Involuntary context switches: 654
	Swaps: 0
	File system inputs: 176
	File system outputs: 5500888
	Socket messages sent: 0
	Socket messages received: 0
	Signals delivered: 0
	Page size (bytes): 4096
	Exit status: 0
```
3. `uring` write with `buffered` IO (48 threads)
```
Run with concurrency = 48, store to disk = 1, direct = 0, use uring = 1
 : 1779.148543 ms
	Command being timed: "./iobench -c 0 -f /run/media/xfs/out.tif"
	User time (seconds): 2.05
	System time (seconds): 31.08
	Percent of CPU this job got: 1855%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 0:01.78
	Average shared text size (kbytes): 0
	Average unshared data size (kbytes): 0
	Average stack size (kbytes): 0
	Average total size (kbytes): 0
	Maximum resident set size (kbytes): 2401952
	Average resident set size (kbytes): 0
	Major (requiring I/O) page faults: 0
	Minor (reclaiming a frame) page faults: 601281
	Voluntary context switches: 47107
	Involuntary context switches: 2326
	Swaps: 0
	File system inputs: 0
	File system outputs: 5500880
	Socket messages sent: 0
	Socket messages received: 0
	Signals delivered: 0
	Page size (bytes): 4096
	Exit status: 0
```

## SYNCHRONOUS

1. synchronous write with `direct` IO (2 threads)
```
Run with concurrency = 2, store to disk = 1, direct = 1, use uring = 0
 : 996.828683 ms
	Command being timed: "./iobench -c 2 -s -d -f /run/media/xfs/io_out.tif"
	User time (seconds): 0.93
	System time (seconds): 0.23
	Percent of CPU this job got: 115%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 0:01.01
	Average shared text size (kbytes): 0
	Average unshared data size (kbytes): 0
	Average stack size (kbytes): 0
	Average total size (kbytes): 0
	Maximum resident set size (kbytes): 52896
	Average resident set size (kbytes): 0
	Major (requiring I/O) page faults: 0
	Minor (reclaiming a frame) page faults: 12287
	Voluntary context switches: 1617
	Involuntary context switches: 162
	Swaps: 0
	File system inputs: 33
	File system outputs: 5500888
	Socket messages sent: 0
	Socket messages received: 0
	Signals delivered: 0
	Page size (bytes): 4096
	Exit status: 0
```

2. `synchronous` write with `direct` IO (48 threads)
```
Run with concurrency = 48, store to disk = 1, direct = 1, use uring = 0
 : 945.703310 ms
	Command being timed: "./iobench -c 0 -d -s -f /run/media/xfs/out.tif"
	User time (seconds): 1.92
	System time (seconds): 1.34
	Percent of CPU this job got: 294%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 0:01.11
	Average shared text size (kbytes): 0
	Average unshared data size (kbytes): 0
	Average stack size (kbytes): 0
	Average total size (kbytes): 0
	Maximum resident set size (kbytes): 181516
	Average resident set size (kbytes): 0
	Major (requiring I/O) page faults: 0
	Minor (reclaiming a frame) page faults: 44475
	Voluntary context switches: 7786
	Involuntary context switches: 216
	Swaps: 0
	File system inputs: 8
	File system outputs: 5500888
	Socket messages sent: 0
	Socket messages received: 0
	Signals delivered: 0
	Page size (bytes): 4096
	Exit status: 0
```
3. `synchronous` write with `buffered` IO (48 threads)

```
Run with concurrency = 48, store to disk = 1, direct = 0, use uring = 0
 : 1713.267466 ms
	Command being timed: "./iobench -c 0 -s -f /run/media/xfs/out.tif"
	User time (seconds): 1.61
	System time (seconds): 30.54
	Percent of CPU this job got: 1870%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 0:01.71
	Average shared text size (kbytes): 0
	Average unshared data size (kbytes): 0
	Average stack size (kbytes): 0
	Average total size (kbytes): 0
	Maximum resident set size (kbytes): 131544
	Average resident set size (kbytes): 0
	Major (requiring I/O) page faults: 0
	Minor (reclaiming a frame) page faults: 33527
	Voluntary context switches: 14663
	Involuntary context switches: 75
	Swaps: 0
	File system inputs: 0
	File system outputs: 5505520
	Socket messages sent: 0
	Socket messages received: 0
	Signals delivered: 0
	Page size (bytes): 4096
	Exit status: 0
```

**Conclusions** 

1. For best performance, use direct IO.
2.  With direct IO using 2 threads, `uring` write is ~20% faster than synchronous write while using about the same number of CPU cycles.

**Comment by boxerab** [comment]
related issue:  https://github.com/axboe/liburing/issues/511

**Comment by HowHsu** [comment]
May I get more details:
1. Is it buffered/direct write
2.  Do you observe the cpu usage when it goes up from 4 therads to 32
3. for the synchronous writes program, does the same problem exist?(the write itself may be the bottleneck)

**Comment by boxerab** [comment]
> May I get more details:
> 
> 1. Is it buffered/direct write
> 2. Do you observe the cpu usage when it goes up from 4 therads to 32
> 3. for the synchronous writes program, does the same problem exist?(the write itself may be the bottleneck)

Thanks. This is for buffered writes. I haven't measured CPU usage but I will do this. 
I've posted the benchmark output above - at highest thread concurrency uring is faster than synchronous,
but it doesn't scale as threads increase, which is what I expected.

**Comment by boxerab** [comment]
@HowHsu I have added CPU usage measurements above.

**Comment by axboe** [comment]
How many files are being written in parallel? There's work in progress for supporting more efficient buffered writes, but in terms of the current situation, io_uring will punt the write to io-wq and io-wq will serialize writes to a single file. That reduces concurrency to the file, but it's a necessary evil because otherwise you just end up with lots of io-wq threads that block on the file lock trying to write to it. This is most likely what you are observing as concurrency being lower than you expect.

Haven't looked at your results yet, but thanks for making a benchmark for this. Will definitely ease analyzing it and figuring out if we can improve things.

**Comment by boxerab** [comment]
@axboe thanks, good to know. My use case is for writes to a single file. When you mention `file lock`, do you mean a lock that the kernel takes when writing to the file ? Or do you mean my user space lock ? Because I realized there is a way of removing my lock if I have multiple rings.

**Comment by axboe** [comment]
I haven't looked at your benchmark yet, was referring to the file system in the kernel locking the file for write.

But in general, NEVER share a ring if you can avoid it. Rings are meant to be cheap enough to use per-thread, and if you do that, then you don't need any synchronization for submitting/completing IO.

For this case, you'd want to share the io-wq backend between rings though, which can be done with ```IORING_SETUP_ATTACH_WQ``` when setting up subsequent rings. That will still serialize writes across the rings for the same file, if you skipped that flag and had multiple rings, then you'd get higher concurrency but also higher locking contention on the file. Something you can experiment with.

**Comment by boxerab** [comment]
Thanks! I will look into multiple rings. And I do think O_DIRECT works for my use case - going to do some experimentation.

**Comment by axboe** [comment]
Stefan has a set of patches for buffered file writes, would be great if we could test those too. I can probably do that, since you posted the test tool. We just need the patches posted publicly for review...

**Comment by boxerab** [comment]
@axboe fantastic! let me know if I can help in any way re testing

**Comment by boxerab** [comment]
@axboe does O_DIRECT also have similar efficiency improvements on BTRFS as well as XFS ? as this is Fedora default.

**Comment by axboe** [comment]
I think btrfs should be fine. In later kernels it uses the iomap infrastructure for O_DIRECT like xfs does.

**Comment by boxerab** [comment]
great, thanks!

**Comment by shrfb** [comment]
How do you invoke your test program? Is it required to pass parameters?

**Comment by boxerab** [comment]
> How do you invoke your test program? Is it required to pass parameters?

if no parameters,then it will iterate through all concurrency levels from 2 to `std::thread::hardware_concurrency()`,
or you can pass in the concurrency you want.

**Comment by shrfb** [comment]
I might miss something, but if I understand the benchmark correctly, then:

- the run procedure in uring_mt_test creates several tasks.
- Each task is defined with a lambda that calls the function tiffFormat.encodePixel().
- The bulk of the work is done in the function  tiffFormat.encodePixels().
- In the function encodePixels() we first take a mutex, the mutex is only released at the end of the function.
  - Basically each task is processed synchronously
  - There might be some I/O parallelism for each task, but
  - we have to wait for the slowest I/O to complete for each task.

Is this assessment correct?

**Comment by boxerab** [comment]
@shrfb that's basically correct, except that most of the work is done in the lambda method before calling encodePixels.
And, with uring, the requests go into the ring and can be executed asynchronously.

The mutex is there because libtiff is not thread safe, so only one thread can "write" to libtiff at a time. The writes to libtiff are just advancing the library offsets, but the actual write is done outside of libtiff, using uring.

I will be hopefully be removing this mutex and using multiple rings.

**Comment by boxerab** [comment]
I found a bug in my benchmark code: I've updated the timings above, this time for a 6 core laptop. With the bug fixed, the uring timing is equal to or greater than the synchronous timing.

**Comment by axboe** [comment]
> With the bug fixed, the uring timing is equal to or greater than the synchronous timing.

Can you clarify what you mean here? I can interpret this in two conflicting ways :-)

**Comment by boxerab** [comment]
> > With the bug fixed, the uring timing is equal to or greater than the synchronous timing.
> 
> Can you clarify what you mean here? I can interpret this in two conflicting ways :-)

No problem : for the 6 core laptop with 5.11 and ext4, uring takes twice as long as synchronous at full concurrency.
For the 24 core desktop with 5.16, btrfs and PCI4 nvme SSD, uring takes ~15% less time than synchronous at full concurrency.

**Comment by axboe** [comment]
That's a big gap on the laptop. Granted buffered writes aren't as efficient as they could be yet (see mention of that patchset from @shrfb, that will fix that gap, but twice as long seems... Wrong. I wonder if 5.16 will do better on the laptop, or if there's something else going on here.

**Comment by boxerab** [comment]
@axboe woops, I got the specs reversed : laptop has 5.17 and btrfs. desktop has 5.11 and ext4. But, yeah, twice is long seems too big a gap - I will do some double checking ono the code.

**Comment by boxerab** [comment]
Here's another interesting data point: two different runs on the laptop at concurrency 6 (there are 6 cores).  For one run, uring timing matched synchronous, while for  the other, it was 3X slower than synchronous !

```
Run with concurrency = 6, store to disk = 0, use uring = 0
 : 667.164988 ms
Run with concurrency = 6, store to disk = 1, use uring = 0
 : 1333.794091 ms
Run with concurrency = 6, store to disk = 1, use uring = 1
 : 3091.314196 ms
 
 \\\\\\\\\\\\\
Run with concurrency = 6, store to disk = 0, use uring = 0
 : 676.898568 ms
Run with concurrency = 6, store to disk = 1, use uring = 0
 : 1303.279492 ms
Run with concurrency = 6, store to disk = 1, use uring = 1
 : 1384.872270 ms
```

**Comment by boxerab** [comment]
@axboe I've fixed some bugs in the benchmark, and updated the timings for desktop and laptop.
I also managed to remove the single locked ring - now there is one ring per thread.
Timings for both synch and asynch are not roughly the same.
I also experimented with 'O_DIRECT" mode on Fedora system, but I found that both synchronous and asynchronous writes
got slower.

**Comment by boxerab** [comment]
I have enabled work queue sharing across threads. This errors out on 5.11, but running a 5.17 VM on my desktop 48 thread machine, I can now almost consistently get asynch write around 10-15% faster than synchronous write for max concurrency. This looks pretty good to me, and once I get `O_DIRECT` enabled, it should improve the situation.

**Comment by boxerab** [comment]
Another interesting data point: synch and asynch take almost the same time, but asynch uses
around 2X the cpu usage of synch.

```
[uring_mt_test]$ /usr/bin/time -v ./uring_mt_test 0 0
Run with concurrency = 12, store to disk = 1, use uring = 0
 : 1047.862855 ms
	Command being timed: "./uring_mt_test 0 0"
	User time (seconds): 2.75
	System time (seconds): 2.19
	Percent of CPU this job got: 348%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 0:01.41
	Average shared text size (kbytes): 0
	Average unshared data size (kbytes): 0
	Average stack size (kbytes): 0
	Average total size (kbytes): 0
	Maximum resident set size (kbytes): 2794068
	Average resident set size (kbytes): 0
	Major (requiring I/O) page faults: 0
	Minor (reclaiming a frame) page faults: 697702
	Voluntary context switches: 1987
	Involuntary context switches: 548
	Swaps: 0
	File system inputs: 0
	File system outputs: 5500920
	Socket messages sent: 0
	Socket messages received: 0
	Signals delivered: 0
	Page size (bytes): 4096
	Exit status: 0
[mt_test]$ /usr/bin/time -v ./uring_mt_test 0 1
Run with concurrency = 12, store to disk = 1, use uring = 1
scheduling : 641.092325 ms
flush : 522.053394 ms
	Command being timed: "./uring_mt_test 0 1"
	User time (seconds): 4.24
	System time (seconds): 7.04
	Percent of CPU this job got: 786%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 0:01.43
	Average shared text size (kbytes): 0
	Average unshared data size (kbytes): 0
	Average stack size (kbytes): 0
	Average total size (kbytes): 0
	Maximum resident set size (kbytes): 1504128
	Average resident set size (kbytes): 0
	Major (requiring I/O) page faults: 0
	Minor (reclaiming a frame) page faults: 392360
	Voluntary context switches: 4805
	Involuntary context switches: 1212
	Swaps: 0
	File system inputs: 0
	File system outputs: 5500888
	Socket messages sent: 0
	Socket messages received: 0
	Signals delivered: 0
	Page size (bytes): 4096
	Exit status: 0
```

**Comment by boxerab** [comment]
I have adjusted the benchmark to be a truer apples-to-apples comparison between synchronous writes and uring writes.
We now have uring significantly faster than synchronous on 48 thread system. Most importantly, the time to schedule the writes with uring is a fraction of the time to write synchronously. Next step is to enable `O_DIRECT`.

**Comment by boxerab** [comment]
@axboe preliminary results for `O_DIRECT` show that it is in fact slower for this benchmark. The buffers are all aligned - offset and length are 512-byte aligned. For direct writes, should the work queue no longer be shared by the rings ?

**Comment by boxerab** [comment]
This benchmark is basically complete. It now supports synchronous and asynchronous writes, using both buffered and direct IO.
Testing on writing a 2.8 GB file show that

1. synchronous and asynchronous takes around the same total time
2.  direct IO is **slower** than buffered
3. using a shared work queue makes no difference to performance

conclusion 2 was surprising. I'm interested to hear of any other ways I can improve performance.

**Comment by boxerab** [comment]
A few more interesting data points for uring writes (from fedora 36 / kernel 5.17 system) :

1. direct IO uses ~20% more wall-clock time than buffered
2. direct IO uses ~50% less CPU than buffered

**Comment by isilence** [comment]
> This benchmark is basically complete. It now supports synchronous and asynchronous writes, using both buffered and direct IO. Testing on writing a 2.8 GB file show that
> 
>     1. synchronous and asynchronous takes around the same total time
> 
>     2. direct IO is **slower** than buffered

Do you flush it to the drive? If the file (or most of it) fits into RAM it'd compare writing to RAM vs actual disk I/O, which is more or less useless.

Also, let me ask what was the I/O size?

**Comment by boxerab** [comment]
> > This benchmark is basically complete. It now supports synchronous and asynchronous writes, using both buffered and direct IO. Testing on writing a 2.8 GB file show that
> > ```
> > 1. synchronous and asynchronous takes around the same total time
> > 
> > 2. direct IO is **slower** than buffered
> > ```
> 
> Do you flush it to the drive? If the file (or most of it) fits into RAM it'd compare writing to RAM vs actual disk I/O, which is more or less useless.
> 
> Also, let me ask what was the I/O size?

Thanks @isilence . Since I close the file descriptor before measuring the timing, there is a flush to disk.

In terms of I/O size, the buffers are aligned on 512 byte boundaries and their size is 64*1024

**Comment by isilence** [comment]
> > > This benchmark is basically complete. It now supports synchronous and asynchronous writes, using both buffered and direct IO. Testing on writing a 2.8 GB file show that
> > > ```
> > > 1. synchronous and asynchronous takes around the same total time
> > > 
> > > 2. direct IO is **slower** than buffered
> > > ```
> > 
> > 
> > Do you flush it to the drive? If the file (or most of it) fits into RAM it'd compare writing to RAM vs actual disk I/O, which is more or less useless.
> > Also, let me ask what was the I/O size?
> 
> Thanks @isilence . Since I close the file descriptor before measuring the timing, there is a flush to disk.

IIRC, close() doesn't synchronously flush it, so I doubt it.

> In terms of I/O size, the buffers are aligned on 512 byte boundaries and their size is 64*1024. The final 64K buffer is usually not fully written, so this would be a buffered write even with `O_DIRECT`, but the rest of the buffers will use direct.

Got it

**Comment by axboe** [comment]
Yes, close(2) gives you no guarantees here. You'd need an fsync() for comparable timing.

**Comment by boxerab** [comment]
> IIRC, close() doesn't synchronously flush it, so I doubt it.

thanks, I've added `fsync` and updated benchmarks. Interesting - currently the best overall performance on 12-thread system
uses :  `uring` / 2 threads / buffered IO

**Comment by boxerab** [comment]
I have updated my benchmarks to show timing for both direct and buffered writes, using synchronous and uring.
Also, my threadripper system was upgraded to 5.17 kernel.  For this system at high concurrency, the fastest timing
is for synchronous buffered writes.

@axboe the improvements you mentioned for uring with buffered io - will those make it into 5.19 ?

**Comment by axboe** [comment]
> the improvements you mentioned for uring with buffered io - will those make it into 5.19 ?

No, they are targeted for 5.20, the latest posting is here:

https://lore.kernel.org/io-uring/20220520183646.2002023-1-shr@fb.com/

**Comment by boxerab** [comment]
> No, they are targeted for 5.20, the latest posting is here:
> 
> https://lore.kernel.org/io-uring/20220520183646.2002023-1-shr@fb.com/

Thanks, very interesting. I see that the patch explicitly mentions XFS. 
Do you ever run benchmarks on other file systems?

**Comment by shrfb** [comment]
Async buffered writes require support from the filesystem. The above mentioned patch provides this for the xfs filesystem. What filesystem are you interested in?

**Comment by boxerab** [comment]
> What filesystem are you interested in?

Currently my test systems are both using btrfs

**Comment by boxerab** [comment]
> After xfs support is upstreamed, the plan is to also add support for Btrfs.

excellent, looking forward to that.

**Comment by boxerab** [comment]
@axboe the current benchmark results show uring with direct io is slower than uring with buffered io, on `btrfs` file system.
Do you have any ideas on why ?  One difference between how the benchmark works for buffered and direct is:

1. for buffered io, the image is broken up into N slices, where each slice is ~2 MB in size. 

2. for direct io, each of these 2 MB slices is broken up further into aligned 32K chunks. So there 
are more individual buffers to schedule for direct. However, these chunks are all scheduled in one
uring call.

**Comment by boxerab** [comment]
I finally got this benchmark working on XFS partition, and I can finally see a huge improvement with
uring + direct vs uring + buffered IO.

uring direct is almost **twice as fast** as uring buffered.

**Comment by boxerab** [comment]
@axboe best results on my system with `XFS` fs and `O_DIRECT`, for single file write, is that `uring` is around 20% faster than synchronous, while using around the same percentage of CPU (concurrency set to 2 threads).  Does that seem reasonable?
I was expecting faster results for `uring`, but maybe I'm not configuring the library properly.
